pipelines/bedbuncher.py

```diff
@@ -77,115 +84,186 @@ def get_bedset_digest(sr):
+    m.update(";".join(sorted([f["md5sum"] for f in sr])).encode('utf-8'))
     return m.hexdigest()
 
 
-def main():
-    pm = pypiper.PipelineManager(name="bedbuncher", outfolder=logs_dir, args=args)
+def mk_file_type(pth, title, md5sum):
+    """
+
+    :param pth:
+    :param title:
+    :return:
+    """
+    rel_to = os.path.abspath(os.path.join(bbc.get_bedbuncher_output_path(), "..", ".."))
+    return {"path": os.path.relpath(pth, rel_to), "title": title}
 
-    # Establish Elasticsearch connection and check status using bbconf
-    bbc.establish_elasticsearch_connection()
 
-    # Use bbconf method to look for files in the ES index
-    q = JSON_to_dict(args.JSON_query_path)
-    search_results = bbc.search_bedfiles(query=q)
+def main():
+    pm = pypiper.PipelineManager(name="bedbuncher", outfolder=logs_dir, args=args)
+    # Use bbconf method to look for files in the database
+    search_results = bbc.bed.select(
+        condition=args.query, condition_val=[args.query_val])
     nhits = len(search_results)
-    if nhits < 1:
-        raise BedBaseConfError("No BED files match the query: {}".format(q))
-    print("{} BED files match the query".format(nhits))
-    hit_ids = {i[JSON_ID_KEY][0]: i[JSON_MD5SUM_KEY][0] for i in search_results}
+    if nhits < 2:
+        raise BedBaseConfError(f"{nhits} BED files match the query: {args.query}")
+    pm.info(f"{nhits} BED files match the query")
+    hit_ids = [i["id"] for i in search_results]
     bedset_digest = get_bedset_digest(search_results)
-    print("bedset digest: {}".format(bedset_digest))
+    pm.info(f"bedset digest: {bedset_digest}")
 
     # check for presence of the output folder and create it if needed
-    output_folder = os.path.abspath(os.path.join(
-        bbc[CFG_PATH_KEY][CFG_BEDBUNCHER_OUTPUT_KEY], bedset_digest))
+    output_folder = os.path.abspath(
+        os.path.join(bbc.get_bedbuncher_output_path(), bedset_digest))
     if not os.path.exists(output_folder):
-        print("Output directory does not exist. Creating: {}".format(output_folder))
+        pm.info(f"Output directory does not exist. Creating: {output_folder}")
         os.makedirs(output_folder)
 
+    # Create TrackHub directory
+    pm.info(f"Creating TrackHub directory for {args.bedset_name}")
+    hub_folder = os.path.join(output_folder, 'bedsetHub')
+    if not os.path.exists(hub_folder):
+        os.makedirs(hub_folder)
+    # write hub.txt file
+    hub_txt = {'hub': 'BEDBASE_'+args.bedset_name,
+            'shortLabel': 'BEDBASE_'+args.bedset_name,
+            'longLabel': args.bedset_name + ' signal tracks',
+            'genomesFile': "genomes.txt",
+            'email': 'bx2ur@virginia.edu',
+            'descriptionUrl': 'http://dev1.bedbase.org/'
+            }
+    f = open(os.path.join(hub_folder, "hub.txt"),"w")
+    f.writelines('{}\t{}\n'.format(k,v) for k, v in hub_txt.items()) 
+    f.close()
+    # write genomes.txt and trackDb.txt
+    genome = []
```

Comment:
I had it an array because I'm not sure, in the future,  if we are going to have more than one genome assembly type in the database.  if the bed files matches the query are from more than one genome assembly, then it will print a message and the trackhub files won't be generated.

Reply:
I haven't tested it but based on what I see here I think it just prints the message and creates `trackDb.txt` for every genome that has been added to `genomes` list.

That's what my next comment is about

Reply:
And since we decided that there are no cross-assembly bedsets (correct me if I'm wrong), list is not needed as we want it to consist of just one element anyway right?

Reply:
> And since we decided that there are no cross-assembly bedsets (correct me if I'm wrong)

you are right. 

But what if the database contains bed files of different assemblies? Then we need to make it required to specify the `genome`  in the query table.  

> I haven't tested it but based on what I see here I think it just prints the message and creates trackDb.txt for every genome that has been added to genomes list.

I intended for it to terminate the pipeline, if there are more than one genome in the list.  should I add `sys.exit(1)` ? and is there something other than `pm.info` for warning/error ?

Reply:
> and is there something other than pm.info for warning/error ?

probably. Check the docs

> Then we need to make it required to specify the genome in the query table.

I think that's what we should do

Reply:
how should I make `query: genome=%s` as required? (or `query: "other->>' genome'=%s"` , since genome is in `other` for now )

Reply:
I don't think you can easily enforce the value the user passes to the pipeline as the query.

just add another _required_ argument `--genome` and compose a query that includes the provided value

alternatively you can not update the query and check if each retrieved record matched the specified genome, but that would be less performant, obviously

Reply:
> just add another _required_ argument `--genome` and compose a query that includes the provided value

For the `query_val`, I can just append the `genome` to the list. What about the `query`?  would the format be something like `"gc_content>%s AND  other->>' genome'=%s"`

Reply:
I'm not sure, just test it. That's what I'd need to do to answer this question :)

Reply:
I removed anything related to the genome list I had before; added a required `--genome` arg; updated the pipeline_interface.yaml.  

I tested the code locally and it created the BED set correctly.